### PR TITLE
codegen: Resolve overload ambiguities created by defaulted arguments

### LINKED ISSUE #43945: caffe2/aten bridge doesn't handle overloads with ambiguous defaulting
The bridge infra generated by [gen_op.py](https://github.com/pytorch/pytorch/blob/master/caffe2/contrib/aten/gen_op.py) isn't designed to handle calls to multiple op overloads whose defaulting may produce ambiguous calls. For example, the two overloads
```
- func: ambig.a(Tensor self, int? i=0) -> Tensor

- func: ambig.b(Tensor self, bool? b=False) -> Tensor
```
will produce (among others) a binding that contains the C++ call `ambig(self)`, resulting in a C++ compiler error. 

This can prevent intuitive overloading schemes from working (see e.g. #42190). It should be possible to disambiguate by supplying default arguments explicitly. 

cc @ezyang @bhosmer @smessmer @ljk53

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45668 Use asymmetric mkldnn convolutions for padding='same' convolutions
* #45667 Add padding='same' mode to conv{1,2,3}d
* **#45666 codegen: Resolve overload ambiguities created by defaulted arguments**

Fixes #43945

### CODE REVIEW

**Comment by ezyang** [comment]
I'm having a hard time understanding what this PR is doing at a high level

**Comment by ezyang** [comment]
Warning: this PR makes ordering of entries matter in `native_functions.yaml`. This was definitely not true for `tools.codegen` codegen, and I kind of suspect it wasn't true for `tools.autograd`, see `sort_declarations` where we explicitly resort for overload resolution. That logic needs to make its way to `tools.codegen`. (This should be especially relevant to you, @ljk53)

**Comment by peterbell10** [comment]
Consider these two different `conv1d` signatures:
```cpp
conv1d(Tensor self, Tensor kernel, Tensor? bias=None, int[1] stride=1, int[1] padding=0, /*etc.*/) -> Tensor
conv1d.padding_mode(Tensor self, Tensor kernel, Tensor? bias=None, int[1] stride=1, str padding="valid", /*etc.*/) -> Tensor
```

If you call `conv1d` without any optional arguments, the python API will select the first overload because it does a linear search. However, the C++ API will give a compiler error because the two overloads are ambiguous. This PR resolves the ambiguity by removing the C++ API defaults for `str padding` and everything to the left of it. So, where previously it was ambiguous, now the compiler will always pick the first overload. Same as in python.

Note: We can't resolve this by removing the defaults from the JIT schema, or python calls like `torch.conv1d(x, kernel, padding="same")` wouldn't be allowed.

**Comment by peterbell10** [comment]
As for ordering: in principle, any ambiguous overloads would need to be functionally equivalent for this to make sense. So, at least in principle, the linear ordering is just an implementation detail.

I would mean you could do something like
```
func.scalar(Tensor self, Scalar? opt=None)
func(Tensor self, Tensor? opt=None)
```
and calling `func(x)` in python calls the `Tensor` version by `sort_declarations` rules, but in C++ it would call the `Scalar` version.

>That logic needs to make its way to `tools.codegen`.

So do you want the sorting done once in `tools/codegen` and have the order in `Declarations.yaml` matter?

**Comment by peterbell10** [comment]
Updated to resolve merge-conflicts with #44087.

**Comment by ezyang** [comment]
FYI, I'm going to fold a version of your `unambiguous_defaults` function into a different WIP refactor which is handling some other cases where we need to disambiguate signatures.

**NO ACTION NEEDED HERE.** However, I thought I'd share a few thoughts I had while adapting your algorithm:

1. I think I'm going to want the algorithm to error in more situations than you have. For example, suppose that you have `f(int a)` taking precedence over `f(int a = 2, int b = 3)`. Your algorithm will strip defaults on the second function to `f(int a, int b)` which indeed solves the ambiguity problem, but also removes `f()` as a valid signature for the combination of overloads. The bug, of course, is that there was a problem with `f(int a)` missing a default. We should just error in this situation. I think there are other similar situations; in general, where the argument lists of two overloads are consistent, their defaulting should also be consistent.
2. I'm kind of hoping there is a higher level algorithmic formulation of this algorithm. In English, I'm looking for something like "we must strip default arguments until the first disambiguation point, i.e., the argument which unambiguously tells us we are handling the second signature rather than the first." Per (1), I'm not sure it matters what the defaulting status on the higher precedence signature is. So we should ONLY be looking at types.

WDYT

**Comment by peterbell10** [comment]
These are some very insightful points, I'm still thinking this over myself but here are some initial thoughts.

> For example, suppose that you have `f(int a)` taking precedence over `f(int a = 2, int b = 3)`.

I would also wonder about if the two signatures' priorities were reversed. By this PR, the two argument version would keep both defaults and the `f(int)` overload wouldn't actually be callable. That would generate another ambiguous C++ overload so possibly should be an error as well. In this case the problem is there is never a valid disambiguation point because it has fewer arguments.

> Per (1), I'm not sure it matters what the defaulting status on the higher precedence signature is. So we should ONLY be looking at types.

So the reason I check if `b` has a default is illustrated by this example:
```
f(int a, int b, int c)
f(int a=0, int b=0)
```
Without the default check, the disambiguation point is found to be `c` and the second overload get stripped of all its defaults. Really though, there was never any ambiguity at all. This comes back to the point of making the order of overload sets independent. If the first overload was written to match the defaults of the second, as you suggest, then it actually introduces a new ambiguity which cannot be fixed by modifying the second signature alone.

Maybe it would make more sense to inspect the entire overload set and actually try to find the "best fit" for the set of arguments, ignoring order in `native_functions.yaml`. I think in this case it might be as simple as the overload with more arguments has the lower priority. However, that would still need to raise an error for the the case from your first example.

**Comment by ezyang** [comment]
Yes, your examples are also leading me to believe that we should formulate the algorithm in an order independent way.

Taking a step back for a moment, what is the purpose of defaulting? Because we can always eliminate ambiguity by eschewing defaulting altogether and just writing each (non-defaulted) overload separately, as you did in the first version of your PR. We prefer not to do this, because this results in backend writers in having to write a lot of implementations of the function. So it seems there are a few constraints here:

* We have some set of overloads that we want to make visible to the end-user API.
* We want to minimize the number of overloads we have to actually write kernels for.
* Each entry in `native_functions.yaml` represents a single kernel you have to write in the backend.

OK, back to order independence. It seems to me that it would be instructive to think of the set of possible overloads as forming a generalized trie, where each branch on the trie corresponds to an argument type. (This would also probably a more efficient way of implement Python argument parsing, as we don't have to repeatedly attempt to reparse prefixes of an argument list.) So if we have `f(int a, int b, int c)`, `f(int a, Tensor b)` and `f(int a)` we have the trie:

```
       +----int---+--int--+--[kernel 1]
--int--+--Tensor--+--[kernel 2]
       +--[kernel 3]
```

Notice that `f(int, int)` doesn't have a terminal in the tree; it's not allowed, even though `f(int)` is allowed. While non-defaulted function signatures only define a single path in the trie, defaulted function signatures define multiple paths. The coherence conditions are when multiple kernels live in the same path of a trie, they must agree.  So to take my example 

```
f(int a)                // kernel 1
f(int a = 2, int b = 3) // kernel 2

--+--int--+--int--[kernel 2]
  |       +--[kernel 1, kernel 2]
  +--[kernel 2]
```

Looking at the trie, we can see that kernel 1 is redundant because there is no case where kernel 1 is the sole overload, and we should tell the developer that something fishy is going on.

But let's say we wish kernel 1 to be used for `f(int)`. I don't think there is any way to use C++ default arguments to express this. We have no choice but to explode our definition into three separate overloads, `f()`, `f(int)`, and `f(int, int)`, to get the semantics we want. We can only use C++ defaulting when the desired kernel choice is consistent for the terminals of an entire suffix in the trie (doesn't have to be an entire path, since you can just avoid defaulting the prefix).

So this all suggests a fully general and correct algorithm would look this:

* Combine all overloads of a function into a trie, recording which kernel would be used for each path
* Make some decision about which function to use (it only really matters that it is deterministic, as the coherence condition says that any kernel that supports a signature should behave equivalently)
* Simplify the trie by collapsing path suffixes for which terminals are supported and the kernel is consistent into C++ default arguments
* Output a C++ overload for every final path in the tree, defaulting only where simplification above worked.

However, this all seems like a bit overkill. Wonder if there's a simpler thing we can do...

**Comment by peterbell10** [comment]
As you're not including this in #45890, how should I proceed. Would it be worth trying to prototype your trie-based algorithm now, or should I just wait for the other codegen changes to be finished?

**Comment by ezyang** [comment]
I think we should get someone else to weigh in on complexity-power ratio here. @bhosmer you said you had some prior discussions, could you check in on this conversation?

**Comment by peterbell10** [comment]
So, as a quick recap there have been 3 suggestions so far. From simplest to most complex:

#### 1. Non-positional defaults (https://github.com/pytorch/pytorch/issues/43945#issuecomment-684910034).  
Allow defaults to appear to the left of non-defaulted arguments. These defaults can't be called from C++ but can be called from python by calling the non-defaulted argument as a keyword argument.

This allows the operator author to manually disambiguate the overloads. But also breaks from the normal python defaulting rules which could be misused.

#### 2. Remove ambiguous defaults from the left (this PR)
This makes the order of declaration in `native_functions.yaml` matter. The past signatures are recorded and if the current signature creates an ambiguous declaration then defaults are removed from the left until it's no longer ambiguous.

This works for simple cases, but can create situations where a default can't be used at all from C++. So, the the developer has to be conscious of the right order to declare their operator overloads.

#### 3. Trie-based solution (https://github.com/pytorch/pytorch/pull/45666#issuecomment-703959802)
Gather together all overloads of an operator and assemble valid positional argument calls into a trie. Then disambiguate so each leaf can only be implemented by a single operator.

This fixes the order-dependence issues, but will be more complicated to implement.

@ezyang, @bhosmer what are your current preferences from these options. Or, is there a better way of doing this?

**Comment by ezyang** [comment]
I'm going to vote for (1), but with a twist: rather than modifying JIT schema to support non-defaulted after defaulted, just add a new field to the YAML that controls cpp signature generation (maybe something like `cpp_no_default_args: foo, bar`). This should be relatively simple to implement but doesn't require big design changes (making native functions ordered, or changing how JIT schema works).

**Comment by bhosmer** [comment]
@peterbell10 @ezyang 

Of these options I like Ed's (1)-with-a-twist, because it preserves python defaulting behavior, preserves order independence, and (importantly) communicates the wrinkle in C++ API behavior upfront in native_functions.yaml.

But note that since it's ambiguous _call sites_ that are errors - rather than the function definitions themselves - there's an even simpler alternative: 

0. Don't change anything. Let the ambiguous C++ definitions get generated as-is, and just avoid making ambiguous calls to them in our generated code. I believe there's only one place where we actually generate calls that make use of defaulting, in [gen_op.py](https://github.com/pytorch/pytorch/blob/master/caffe2/contrib/aten/gen_op.py), which should be simple to modify to give the default values explicitly.

How we feel about this option probably depends on whether the overlapping C++ functions feel like a ticking API time bomb, or a "don't hold it like that" kind of thing. Like, how bad a user experience will it be to call one of these and get an "ambiguous call site" error? 

Aside: since Ed's solution will also rely on actual call site errors to alert the op writer to ambiguity if they hadn't themselves noticed, it means that we're implicitly relying on gen_op.py as sort of a de facto C++ API coverage test at the moment. 

Aside: long-term it's appealing to move to something like Ed's trie data model, since we could use it to generate multiple artifacts (the C++ overload set, fast overload resolution, etc.)

**Comment by peterbell10** [comment]
Okay, it sounds like (1)-with-a-twist is the winner. 

>Don't change anything. Let the ambiguous C++ definitions get generated as-is, and just avoid making ambiguous calls to them in our generated code.

That would mean introducing a new overload will break compilation, not just of pytorch itself, but also all users of the C++ API. e.g. to add padding modes, `torch::conv2d(a, b)` would fail to compile and have to become
```cpp
torch::conv2d(a, b, /*bias=*/{}, /*stride=*/1, /*padding=*/0)
```

**Comment by bhosmer** [comment]
> That would mean introducing a new overload will break compilation, not just of pytorch itself, but also all users of the C++ API

Ah of course these are incremental additions, not blocks of overloads added all at once. Sorry - forest for the trees :)

**Comment by peterbell10** [comment]
@ezyang PTAL. I put the `cpp_no_default` marker into the `Argument` class so that the c++ generator can do the `Argument -> CppArgument` mapping without needing additional context.

File: tools/codegen/api/cpp.py [hunk 1]
```diff
@@ -232,12 +232,15 @@ def argument_not_this(
...
-            default=default_expr(a.default, a.type) if a.default is not None else None,
+            default=(default_expr(a.default, a.type)
+                     if a.default is not None and not a.cpp_no_default else None),
             argument=a,
         )
     elif isinstance(a, TensorOptionsArguments):
         default = None
-        if all(x.default == "None" for x in a.all()):
+        if all(x.cpp_no_default for x in a.all()):
+            default = None
```
Comment (ezyang) [inline tools/codegen/api/cpp.py:242]: This is kind of weird; probably should raise an error in frontend if you `cpp_no_default` some but not all tensor options arguments.

File: tools/codegen/model.py [hunk 2]
```diff
@@ -153,9 +153,17 @@ def from_yaml(ei: Dict[str, object], loc: 'Location') -> 'NativeFunction':
...
+
         funcs = e.pop('func')
         assert isinstance(funcs, str), f'not a str: {funcs}'
-        func = FunctionSchema.parse(funcs)
+        func = FunctionSchema.parse(funcs, cpp_no_default_args)
+
+        defaulted_arguments = {a.name for a in func.schema_order_arguments()
+                               if a.default is not None}
+        invalid_args = set(cpp_no_default_args).difference(defaulted_arguments)
+        assert len(invalid_args) == 0, f'Invalid cpp_no_default_args: {invalid_args}'
```
Comment (ezyang) [inline tools/codegen/model.py:166]: nit: Put this in `__post_init__`; it doesn't really matter too much the way the code exists today, but the invariant you're testing here would apply if someone manually constructed a NativeFunction, so it shouldn't live in the parsing code.

File: tools/codegen/model.py [hunk 3]
```diff
@@ -746,7 +755,7 @@ class Argument:
     annotation: Optional[Annotation]
 
     @staticmethod
-    def parse(arg: str) -> 'Argument':
+    def parse(arg: str, cpp_no_default_args: List[str]) -> 'Argument':
```
Comment (ezyang) [inline tools/codegen/model.py:758]: nit: Prefer to do the `name in cpp_no_default_args` test in the call site, and just accept a boolean here

**Comment by ezyang** [comment]
Short and sweet.

One thing that I'm a little ambivalent about is that we store the no-cpp-defaultness directly in Argument itself. I agree that this is the most convenient place to put it, but it breaks the invariant that an Argument directly represents the string it was parsed from (for good reason). The other options are to explicitly maintain context in cpp translation, or implicitly maintain context using the locals mechanism.

Do you know how wordy it is to do the explicit context in cpp translation version? I think I'd prefer it, if it wasn't too much code; it maintains the invariant that the structures refer to their parse trees and it also limits the scope of no-cpp-defaultness to the cpp portion.

File: tools/codegen/model.py (hunk 2 above)
Comment (peterbell10) [reply tools/codegen/model.py:166]: The list of `cpp_no_default_args` only exists in the parse function. If you manually construct a NativeFunction, then you would say `Argument(..., cpp_no_default=True)` instead.

File: tools/codegen/model.py (hunk 3 above)
Comment (peterbell10) [reply tools/codegen/model.py:758]: But name isn't parsed until inside this function?

**Comment by peterbell10** [comment]
For comparison, I've implemented `cpp_no_default_args` left outside of the arguments struct: https://github.com/pytorch/pytorch/commit/d534f7d4c555a37fd178c143098b8537a5a05d61. It requires plumbing `cpp_no_default_args` through various places in codegen and also a few places in autograd.

**Comment by ezyang** [comment]
Thank you. I see the plumbing difficulties.

One easy way to fix the problem is to change the signature of CppSignatureGroup constructor to take a NativeFunction rather than FunctionSchema, then you don't need to do extra manual plumbing (you still have to modify all the sites, but you don't have to goop `cpp_no_default_args=f.cpp_no_default_args`).

I think what I will probably do is commandeer this patch and reimplement it on top of https://github.com/pytorch/pytorch/pull/49122 which will conflict (won't be too hard to reimplement, I think). Bug me if I fail to follow through on this.

**Comment by peterbell10** [comment]
Superceded by #49348

**Pull request closed without merging**